src/Streamly/Internal/Data/Producer/Type.hs

```diff
@@ -0,0 +1,163 @@
+-- |
+-- Module      : Streamly.Internal.Data.Producer.Type
+-- Copyright   : (c) 2021 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- See "Streamly.Internal.Data.Producer" for introduction.
+--
+
+module Streamly.Internal.Data.Producer.Type
+    (
+    -- * Type
+    Producer (..)
+
+    -- * Producers
+    , nil
+    , nilM
+    , unfoldrM
+    , fromList
+
+    -- * Mapping
+    , translate
+    , lmap
+
+    -- * Nesting
+    , NestedLoop (..)
+    , concat
+    )
+where
+
+#include "inline.hs"
+
+import Fusion.Plugin.Types (Fuse(..))
+import Streamly.Internal.Data.Stream.StreamD.Step (Step(..))
+import Prelude hiding (concat)
+
+------------------------------------------------------------------------------
+-- Type
+------------------------------------------------------------------------------
+
+-- | A @Producer m a b@ is a generator of a stream of values of type @b@ from a
+-- seed of type 'a' in 'Monad' @m@.
+--
+-- /Internal/
+
+data Producer m a b =
+    -- | @Producer step inject extract@
+    forall s. Producer (s -> m (Step s b)) (a -> m s) (s -> m a)
+
+------------------------------------------------------------------------------
+-- Producers
+------------------------------------------------------------------------------
+
+{-# INLINE nilM #-}
+nilM :: Monad m => (a -> m c) -> Producer m a b
+nilM f = Producer step return return
+
+    where
+
+    {-# INLINE_LATE step #-}
+    step x = f x >> return Stop
+
+{-# INLINE nil #-}
+nil :: Monad m => Producer m a b
+nil = nilM (\_ -> return ())
+
+{-# INLINE unfoldrM #-}
+unfoldrM :: Monad m => (a -> m (Maybe (b, a))) -> Producer m a b
+unfoldrM next = Producer step return return
+  where
+    {-# INLINE_LATE step #-}
+    step st = do
+        r <- next st
+        return $ case r of
+            Just (x, s) -> Yield x s
+            Nothing -> Stop
+
+-- | Convert a list of pure values to a 'Stream'
+--
+-- /Internal/
+{-# INLINE_LATE fromList #-}
```

Comment:
Is this the right INLINE phase?

Reply:
This is copied from Unfold, there may have been a reason for this.

Reply:
Keeping it as it is.